### PR TITLE
Rework volume path mapping for robustness

### PR DESCRIPTION
**Pull request by Paliverse**
Supersedes #175 (closing it) — same goal, but rebuilt from a clean base after review feedback
  there surfaced deeper issues in both my patch and the existing implementation.

  ## Defects fixed in `VolumeHelper`

  1. **Volumes without mount points poisoned matching** — they produced empty/garbage
  `DriveLetter` entries; now skipped at enumeration.
  2. **Stale buffer corruption** — the reused 64K buffers were never cleared, so a shorter string
  left residue from the previous volume (`TrimEnd` on the whole buffer kept embedded NULs and
  stale text), silently corrupting mappings. Buffers are now cleared per iteration and returned
  lengths are honored.
  3. **Search handle leak** — `FindVolumeClose` was never called; now closed in a `finally`, and
  `FindNextVolume` failures other than `ERROR_NO_MORE_FILES` are surfaced instead of silently
  truncating the list.
  4. **`UriFormatException` escaping `throwOnError: false`** — `NormalizePath` used `new
  Uri(path)`, which throws on garbage input; replaced with exception-guarded `Path.GetFullPath`.
  5. **Reverse translation rejected non-`HarddiskVolume` devices** — the regex gate meant entries
  on dynamic disks (`\Device\HarddiskDmVolumes\...`) and similar could be *written* to the
  driver but were silently dropped when reading back. Per the QueryDosDevice docs the target form
  isn't guaranteed (Microsoft's own "Displaying Volume Paths" sample output includes
  `\Device\CdRom0`), so matching is now mapping-driven with a boundary check (`HarddiskVolume1`
  can no longer claim `HarddiskVolume10` paths).
  6. **Mounted folders resolved through the wrong volume** — only the first mount point per
  volume was kept and first-match selection could map a file under a mounted folder via the
  volume root, producing entries the driver never matches. Now one mapping per mount point with
  longest-prefix selection, and the forward direction canonicalizes the input once and slices
  that same string (fixes relative, forward-slash and `..` inputs).

  ## New: DefineDosDevice-backed drives

  `FindFirstVolume` only sees Mount Manager volumes, so VeraCrypt/ImDisk-style drives never
  translated. A drive-letter sweep (`QueryDosDevice("X:")`) now picks them up, with guards:
  network-redirector targets (session-specific `\;X:` segments) are excluded, and one level of
  `\??\` indirection (SUBST) resolves through the Mount Manager mappings.

  ## Behavior changes to be aware of

  - Volume **enumeration** failure now throws `Win32Exception` even with `throwOnError: false`
  (previously it degraded to an empty mapping list, which could silently wipe the blocklist
  during an Add/Remove rewrite). `throwOnError` still governs input-level failures, which return
  `null` as before (docs said "empty string" but the code always returned `null`; docs
  corrected).
  - `ApplicationPaths` now returns entries on non-`HarddiskVolume` devices that were previously
  dropped.
  - No public API changes — all new types are `internal`.

  ## Tests

  `VolumeHelper` had no coverage; this adds 17 tests via an `IVolumeMappingProvider` seam
  (`InternalsVisibleTo("Tests")` in the csproj), including longest-match selection,
  prefix-collision boundaries, canonicalization cases, round-tripping, MULTI_SZ parsing, and one
  live enumeration smoke test. Full suite (24 tests, including the existing driver integration
  tests round-tripping through this code against the real driver) passes; `dotnet build -c
  Release` of the solution is clean on all four TFMs.

## Summary by CodeRabbit

* **Refactor**
  * Improved device↔path conversion and mapping for more reliable, canonical resolution of complex paths.
  * Updated service whitelist handling to operate in device-path space for robust add/remove semantics.
* **Tests**
  * Added comprehensive tests covering many path/mapping edge cases, mount-point tie-breaks, and a disposable temp-file helper.
  * Marked some hardware-dependent tests as explicit so they don’t run by default.
* **Chores**
  * CI now runs tests on Windows and includes a dedicated test step; test assembly granted access to internals.
* **Other**
  * Added a missing native API entry used during volume enumeration.

### CODE REVIEW

File: Tests/VolumeHelperTests.cs [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+                Directory.Delete(_directory, recursive: true);
+            }
+
+            try
+            {
+                // Remove the shared parent directory once the last test directory is gone
+                Directory.Delete(System.IO.Path.GetDirectoryName(_directory)!, recursive: false);
+            }
+            catch (IOException)
+            {
```
[the remark this answers is not in the document]
Comment (Paliverse) [reply Tests/VolumeHelperTests.cs:389]: DirectoryNotFoundException derives from IOException, so the existing catch already handles the deleted-parent case. There’s also no parallel execution in this test project (no [Parallelizable], NUnit defaults to sequential), so the race described can’t occur.

p.s from a real human, these bots be wilding. Can’t wait for bot vs bot in arenas

**Comment by nefarius** [comment]
@coderabbitai why are you not using the limits of the Pro Plus plan you should be on?

**Comment by nefarius** [comment]
@coderabbitai review

**Comment by nefarius** [comment]
@Paliverse do you wanna test these changes? I can also merge and make a new pre-release NuGet if required.

**Comment by Paliverse** [comment]
@nefarius I cant test to know if this would resolve the issues that users have, but the intent of this rework was to solve this issue from analytics. Past 30 days, 37 users have gotten this:

```
System.UriFormatException: Invalid URI: The URI scheme is not valid.
   at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind, UriCreationOptions& creationOptions)
   at System.Uri..ctor(String uriString)
   at Nefarius.Drivers.HidHide.Util.VolumeHelper.NormalizePath(String path)
   at Nefarius.Drivers.HidHide.Util.VolumeHelper.<>c__DisplayClass7_0.<PathToDosDevicePath>b__0(VolumeMeta m)
   at System.Linq.Enumerable.TryGetSingle[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)
   at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
   at Nefarius.Drivers.HidHide.Util.VolumeHelper.PathToDosDevicePath(String path, Boolean throwOnError)
   at Nefarius.Drivers.HidHide.HidHideControlService.<AddApplicationPath>b__39_0(String p)
   at System.Linq.Enumerable.IteratorSelectIterator`2.MoveNext()
   at System.Linq.Enumerable.IEnumerableWhereIterator`1.MoveNext()
   at System.Linq.Enumerable.Aggregate[TSource,TAccumulate](IEnumerable`1 source, TAccumulate seed, Func`3 func)
   at Nefarius.Drivers.HidHide.Util.MultiSzHelper.StringArrayToMultiSzPointer(IEnumerable`1 instances, Int32& length)
   at Nefarius.Drivers.HidHide.HidHideControlService.AddApplicationPath(String path, Boolean throwIfInvalid)
   at DSX.Core.DeviceManagement.DeviceManager.AddDSXPathToHidHide()
```

So we would only know if it's fixed if these events are no longer being reported to AppInsights.

I've also figured out a way to know which applications have opened handles to x device via the instance path without needing admin that can potentially be integrated into this API if you're interested in it.
